config/config.go

```diff
@@ -737,38 +745,71 @@ func fetchConfig(u *url.URL) ([]byte, error) {
 	req.Header.Add("Accept", "application/toml")
 	req.Header.Set("User-Agent", internal.ProductToken())
 
-	retries := 3
-	for i := 0; i <= retries; i++ {
-		body, err, retry := func() ([]byte, error, bool) {
-			resp, err := http.DefaultClient.Do(req)
-			if err != nil {
-				return nil, fmt.Errorf("retry %d of %d failed connecting to HTTP config server: %w", i, retries, err), false
-			}
-			defer resp.Body.Close()
-			if resp.StatusCode != http.StatusOK {
-				if i < retries {
-					log.Printf("Error getting HTTP config.  Retry %d of %d in %s.  Status=%d", i, retries, httpLoadConfigRetryInterval, resp.StatusCode)
-					return nil, nil, true
-				}
-				return nil, fmt.Errorf("retry %d of %d failed to retrieve remote config: %s", i, retries, resp.Status), false
-			}
-			body, err := io.ReadAll(resp.Body)
-			return body, err, false
-		}()
+	rawQuery, err := AddHostParam(req.URL)
+	if err != nil {
+		return nil, err
+	}
+	req.URL.RawQuery = rawQuery
 
-		if err != nil {
-			return nil, err
+	var totalAttempts int
+	if urlRetryAttempts == -1 {
+		totalAttempts = -1
+		log.Printf("Using unlimited number of attempts to fetch HTTP config")
+	} else if urlRetryAttempts == 0 {
+		totalAttempts = 3
+	} else if urlRetryAttempts > 0 {
+		totalAttempts = urlRetryAttempts
+	} else {
+		return nil, fmt.Errorf("invalid number of attempts: %d", urlRetryAttempts)
+	}
+
+	attempt := 0
+	for {
+		body, err := requestURLConfig(req)
+		if err == nil {
+			return body, nil
 		}
 
-		if retry {
-			time.Sleep(httpLoadConfigRetryInterval)
-			continue
+		log.Printf("Error getting HTTP config (attempt %d of %d): %s", attempt, totalAttempts, err)
+		if urlRetryAttempts != -1 && attempt >= totalAttempts {
+			return nil, err
 		}
 
-		return body, err
+		time.Sleep(httpLoadConfigRetryInterval)
+		attempt++
+	}
+}
+
+func AddHostParam(url *url.URL) (string, error) {
+
+	val := url.Query()
+
+	host, err := os.Hostname()
+	if err != nil {
+		return "", err
+	}
+
+    val.Add("host", host)
+    return val.Encode(), nil
+}
+
+func requestURLConfig(req *http.Request) ([]byte, error) {
+	resp, err := http.DefaultClient.Do(req)
```

Comment:
Using default http client - bad idea. There is no timeouts set

Reply:
Yes, but in this particular case it works correctly because we have manual retries with 10 second delays.
I would suggest to rewrite it, but in next iterations

Reply:
https://github.com/devopsext/telegraf/pull/138/files#diff-fe44f09c4d5977b5f5eaea29170b6a0748819c9d02271746a20d81a5f3efca17R778